segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,377 @@
+                while (length != 0) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                    offset += bytesWritten;
+                    length -= bytesWritten;
+                }
+            }
+            return null;
+        }
+    }
+
+    private boolean isWritableFile(Path path) throws IOException {
+        PosixFileAttributes attrs = null;
+            attrs = Files.readAttributes(path,
+                    PosixFileAttributes.class);
+            return attrs.permissions().contains(OWNER_WRITE);
+    }
+
+    @SneakyThrows
+    private Void syncSeal(SegmentHandle handle) {
+
+        if (handle.isReadOnly()) {
+            log.info("Seal called on a read handle for segment {}", handle.getSegmentName());
+            throw new IllegalArgumentException(handle.getSegmentName());
+        }
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as r--r--r-- 444
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            Files.setPosixFilePermissions(Paths.get(config.getFilesystemRoot(), handle.getSegmentName()), perms);
+            log.info("Successfully sealed segment {}", handle.getSegmentName());
+            return null;
+    }
+
+    @SneakyThrows(IOException.class)
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment) {
+
+        Path sourcePath = Paths.get(config.getFilesystemRoot(), sourceSegment);
+        Path targetPath = Paths.get(config.getFilesystemRoot(), targetHandle.getSegmentName());
+
+        try (FileChannel targetChannel = (FileChannel) Files.newByteChannel(targetPath, EnumSet.of(StandardOpenOption.APPEND));
+             RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+            if (isWritableFile(sourcePath)) {
+                throw new IllegalStateException(sourceSegment);
+            }
+
+            targetChannel.transferFrom(sourceFile.getChannel(), offset, sourceFile.length());
+            Files.delete(Paths.get(config.getFilesystemRoot(), sourceSegment));
+            return null;
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private Void syncDelete(SegmentHandle handle) {
+            Files.delete(Paths.get(config.getFilesystemRoot(), handle.getSegmentName()));
+            return null;
+    }
+
+    /**
+     * Executes the given supplier asynchronously and returns a Future that will be completed with the result.
+     */
+    private <R> CompletableFuture<R> supplyAsync(String segmentName, Supplier<R> operation) {
+        CompletableFuture<R> result = new CompletableFuture<>();
+        this.executor.execute(() -> {
+            try {
+                result.complete(operation.get());
+            } catch (Exception e) {
```

Comment:
`catch Throwable`. If you stumble across some assertion error, your future will be left hanging forever.